### PR TITLE
`Image` refactoring and fixes to `ExposedPorts` and `Labels` logic

### PR DESCRIPTION
**Pull request by vlada-shubina**
Fixes https://github.com/dotnet/sdk-container-builds/issues/332

Also:
- split `Image` class into several
- split logic working with json configuration to separate class
- implemented builder pattern for cleaner code and performance - serialization of updated image is done once in the end
- some small perf improvements: calculating image size, digest, sha only once.

Open items: 
- [x] unit tests testing the fix
- [x] clarify expected behavior with @baronfel

### CODE REVIEW

File: Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+                }
+                """;
+
+            JsonNode? node = JsonNode.Parse(simpleImageConfig);
+            Assert.IsNotNull(node);
+
+            ImageConfig baseConfig = new ImageConfig(node);
+
+            baseConfig.AddLabel("testLabel1", "v1");
+            baseConfig.AddLabel("existing2", "v2");
```
Comment (vlada-shubina) [inline Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs:138]: @baronfel  is it a valid scenario? what should be the result in this case?

File: Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+            JsonNode? result = JsonNode.Parse(readyImage);
+
+            var resultPorts = result?["config"]?["ExposedPorts"] as JsonObject;
+            Assert.IsNotNull(resultPorts);
+
+            // Assert.AreEqual(3, resultPorts.Count);   --> 4 6100/tcp also exists
+            Assert.IsNotNull(resultPorts["6000/tcp"] as JsonObject);
+            Assert.IsNotNull(resultPorts["6010/udp"] as JsonObject);
+            Assert.IsNotNull(resultPorts["6100/udp"] as JsonObject);
```
Comment (vlada-shubina) [inline Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs:291]: @baronfel is it correct behavior? should 6100 be replaced in this case?

File: Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs (hunk 1 above)
Comment (baronfel) [reply Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs:138]: you can overwrite label names, and the new values should be used - similar to updating a dictionary or handling repeated CLI arguments: last one wins.

File: Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs (hunk 2 above)
Comment (baronfel) [reply Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs:291]: the `key` for the port mapping is two parts: required port number and optional port type (`tcp` or `udp`). if no port type is specified than `tcp` is assumed by the engine (not necessarily by us at this layer). In addition, multiple protocols _can_ bind to the same port, so adding a new entry with the same port but different protocol is allowed.

So adding a `6100/udp` in this case would represent 4 total port bindings, 600/tcp, 6010/udp, and 6100 on both tcp and udp.

Comment (vlada-shubina) [reply Test.Microsoft.NET.Build.Containers/ImageBuilderTests.cs:291]: The current implementation assumes that if no port type is specified, the port type is `tcp`.
Assuming replacing `6200` with `6200/tcp` as in this test is ok.

**Review by baronfel — APPROVED** [review]
LGTM but needs to have conflicts resolved.

**Pull request merged**